api/main.py

```diff
@@ -220,7 +220,7 @@ async def get_nodes(request: Request, kind: str = "node",
         )
 
     translated_params = model.translate_fields(query_params)
-    return paginate(await db.find_by_attributes(model, translated_params))
+    return await db.find_by_attributes(model, translated_params)
```

Comment:
The `limit` and `offset` arguments are dropped from `query_params`, so they won't be in `translated_params` either.  So I believe they're not passed to `find_by_attributes()` and pagination is still not happening at the database level.

Are you actually returning chunks of the expected length?  I wouldn't be surprised if the returned data was the whole set with no pagination enabled.

Reply:
Even if we do not provide values of `limit` and `offset` to the method,  `paginate` is able to get those values and uses it to only get the required chunk from the database. So, pagination is working at the db level.

The `paginate` method is fetching the variables using context variable.
Below is the piece of code:
```
@dataclass
class RawParams:
    limit: int
    offset: int


class AbstractParams(ABC):
    @abstractmethod
    def to_raw_params(self) -> RawParams:
        pass
```

I also tried to print the output of `await paginate(collection=col, query_filter=attributes)`. The output has the actual value of `limit` and `offset` passed to the endpoint.

Reply:
So how are the limit and offset values passed to the paginate function if they're not in the parameters?

Reply:
I guess this is some implicit FastAPI feature to provide the query parameters while bypassing the endpoint handler?  It would be good to understand this completely.  Maybe you've already found the documentation for this?

Reply:
I haven't found any related documentation. I studied the `fastapi-pagination` code. As per my understanding below is the scenario here:

In `add_pagination` function, there is `_update_route`. In which query parameters are determined using `_create_params_dependency` and somehow stored in the context variables.
Please check the code here: https://github.com/uriyyo/fastapi-pagination/blob/main/fastapi_pagination/api.py#L161

The parameters are being fetched using `__params_type__` of the mentioned `response_model`. We have `PageModel` in our case.
If you add any test variable in `PageModel.__params_type__`, it will be reflected in the list of parameters on http://localhost:8001/docs#/default/get_nodes_nodes_get
 
So, the /nodes handler function does not need to externally specify `limit` and `offset` in the signature.
`async def get_nodes(request: Request, kind: str = "node"):` would work too.

Reply:
I see.  So maybe there's some potential simplification to be done there.